File: backend/app/models/testing.py
"""Pydantic models for the Bronze testing framework."""

from __future__ import annotations

from datetime import datetime
from typing import List, Optional, Union

from pydantic import BaseModel


class AssertionSpec(BaseModel):
    """SQL-based assertion to run after a test case."""

    type: str  # row_count | scalar_equals | row_count_gte
    sql: str
    expected: Union[int, float, str]
    description: str


class TestCase(BaseModel):
    """A single test scenario within a suite."""

    id: str
    name: str
    category: str  # insert | update | delete | late_arriving | null_pk | duplicate | idempotency | audit
    positive: bool
    data_file: Optional[str] = None
    setup_data_file: Optional[str] = None  # TC004 pattern: upload + run this first
    setup: List[str] = []  # e.g. ["truncate_test_table"]
    teardown: List[str] = []
    assertions: List[AssertionSpec] = []


class TestSuite(BaseModel):
    """Complete test suite definition for a source."""

    source_name: str
    source_type: str
    primary_keys: List[str]
    target_table: str
    test_catalog: str
    test_schema: str
    test_cases: List[TestCase]


class AssertionResult(BaseModel):
    """Result of a single SQL assertion."""

    type: str
    description: str
    expected: Union[int, float, str]
    actual: Optional[Union[int, float, str]] = None
    passed: bool
    sql: str
    error: Optional[str] = None


class TestCaseResult(BaseModel):
    """Result of executing one test case."""

    id: str
    name: str
    category: str
    positive: bool
    status: str  # PASSED | FAILED | RUNNING | SKIPPED | ERROR
    duration_seconds: Optional[float] = None
    assertions: List[AssertionResult] = []
    error: Optional[str] = None


class TestRunSummary(BaseModel):
    """Aggregate counts for a test run."""

    total: int
    passed: int
    failed: int
    skipped: int


class TestRunResult(BaseModel):
    """Full result record for one suite execution."""

    run_id: str
    source_name: str
    started_at: datetime
    completed_at: Optional[datetime] = None
    duration_seconds: Optional[float] = None
    overall_status: str  # PASSED | FAILED | RUNNING
    environment: str
    tester: str = "portal-test-framework"
    summary: TestRunSummary
    test_cases: List[TestCaseResult] = []


class TestSuiteSummary(BaseModel):
    """Lightweight summary for the suite list view."""

    source_name: str
    source_type: str
    primary_keys: List[str]
    target_table: str
    test_count: int
    last_run_status: Optional[str] = None
    last_run_at: Optional[str] = None


class TestSuiteListResponse(BaseModel):
    suites: List[TestSuiteSummary]
    total: int


class TestRunListResponse(BaseModel):
    source_name: str
    runs: List[TestRunResult]
    total: int


class GenerateSuiteResponse(BaseModel):
    source_name: str
    message: str
    test_count: int


class RunSuiteResponse(BaseModel):
    run_id: str
    source_name: str
    message: str


# ── AI test case generator models ─────────────────────────────────────────────

class TcGenerateRequest(BaseModel):
    prompt: str


class TcGeneratePreview(BaseModel):
    tc_id: str
    name: str
    category: str
    positive: bool
    setup: List[str] = []
    teardown: List[str] = []
    assertions: List[AssertionSpec] = []
    data_file_name: str
    data_records: List[dict] = []
    explanation: str


class TcConfirmRequest(BaseModel):
    tc_id: str
    name: str
    category: str
    positive: bool
    setup: List[str] = []
    teardown: List[str] = []
    assertions: List[AssertionSpec] = []
    data_file_name: str
    data_records: List[dict] = []


class TcConfirmResponse(BaseModel):
    tc_id: str
    data_file: str
    message: str
    result: TestCaseResult

File: frontend/src/types/testing.ts
/** TypeScript interfaces mirroring portal/backend/app/models/testing.py */

export type TestStatus =
  | "PASSED"
  | "FAILED"
  | "RUNNING"
  | "SKIPPED"
  | "ERROR"
  | "NOT_RUN";

export type TestCategory =
  | "insert"
  | "update"
  | "delete"
  | "late_arriving"
  | "null_pk"
  | "duplicate"
  | "idempotency"
  | "audit"
  | "data_quality";

export interface AssertionSpec {
  type: string;
  sql: string;
  expected: number | string;
  description: string;
}

export interface TestCase {
  id: string;
  name: string;
  category: TestCategory;
  positive: boolean;
  data_file: string | null;
  setup_data_file: string | null;
  setup: string[];
  teardown: string[];
  assertions: AssertionSpec[];
}

export interface TestSuite {
  source_name: string;
  source_type: string;
  primary_keys: string[];
  target_table: string;
  test_catalog: string;
  test_schema: string;
  test_cases: TestCase[];
}

export interface AssertionResult {
  type: string;
  description: string;
  expected: number | string;
  actual: number | string | null;
  passed: boolean;
  sql: string;
  error?: string | null;
}

export interface TestCaseResult {
  id: string;
  name: string;
  category: TestCategory;
  positive: boolean;
  status: TestStatus;
  duration_seconds: number | null;
  assertions: AssertionResult[];
  error: string | null;
}

export interface TestRunSummary {
  total: number;
  passed: number;
  failed: number;
  skipped: number;
}

export interface TestRunResult {
  run_id: string;
  source_name: string;
  started_at: string;
  completed_at: string | null;
  duration_seconds: number | null;
  overall_status: TestStatus;
  environment: string;
  tester: string;
  summary: TestRunSummary;
  test_cases: TestCaseResult[];
}

export interface TestSuiteSummary {
  source_name: string;
  source_type: string;
  primary_keys: string[];
  target_table: string;
  test_count: number;
  last_run_status: TestStatus | null;
  last_run_at: string | null;
}

export interface TestSuiteListResponse {
  suites: TestSuiteSummary[];
  total: number;
}

export interface TestRunListResponse {
  source_name: string;
  runs: TestRunResult[];
  total: number;
}

export interface GenerateSuiteResponse {
  source_name: string;
  message: string;
  test_count: number;
}

export interface RunSuiteResponse {
  run_id: string;
  source_name: string;
  message: string;
}

// ── AI test case generator types ──────────────────────────────────────────────

export interface TcGeneratePreview {
  tc_id: string;
  name: string;
  category: string;
  positive: boolean;
  setup: string[];
  teardown: string[];
  assertions: AssertionSpec[];
  data_file_name: string;
  data_records: Record<string, unknown>[];
  explanation: string;
}

export interface TcConfirmResponse {
  tc_id: string;
  data_file: string;
  message: string;
  result: TestCaseResult;
}
